File: python/tests/test_player.py
from othello.models import Disk
from othello.player import Player, PlayerType
from othello.settings import PlayerSettings


def test_new_player():
    player = Player(Disk.BLACK)
    assert player._disk == Disk.BLACK
    assert player.human() is True
    assert player.computer() is False
    assert player.can_play is True
    assert player._rounds_played == 0
    assert player._settings == PlayerSettings.default()


def test_reset_player():
    player = Player(Disk.WHITE, PlayerSettings.default())
    player.can_play = False
    player._rounds_played = 10
    player.reset()
    assert player.can_play is True
    assert player._rounds_played == 0


def test_set_human():
    player = Player(Disk.BLACK)
    assert player.human() is True
    assert player.computer() is False

    player.set_computer()
    assert player.computer() is True
    assert player._player_type == PlayerType.COMPUTER

    player.set_human()
    assert player.human() is True
    assert player._player_type == PlayerType.HUMAN


def test_player_type_string():
    player = Player(Disk.BLACK)
    assert player.type_string() == "Human   "

    player.set_computer()
    assert player.type_string() == "Computer"

File: cpp/tests/test_player.cpp
#include "player.hpp"

#include <gtest/gtest.h>

namespace othello
{

/// Test fixture that has access to the private members of `Player`.
class PlayerTest : public ::testing::Test
{
protected:
    static Disk disk(const Player& player)
    {
        return player.disk;
    }

    static int rounds_played(const Player& player)
    {
        return player.rounds_played;
    }

    static void set_rounds_played(Player& player, const int rounds)
    {
        player.rounds_played = rounds;
    }

    static PlayerSettings settings(const Player& player)
    {
        return player.settings;
    }
};

TEST_F(PlayerTest, new_player)
{
    const Player player = Player::black(PlayerSettings {});
    EXPECT_EQ(disk(player), Disk::black);
    EXPECT_TRUE(player.human());
    EXPECT_TRUE(player.can_play);
    EXPECT_EQ(rounds_played(player), 0);
    EXPECT_EQ(settings(player), PlayerSettings {});
}

TEST_F(PlayerTest, reset_player)
{
    Player player = Player::white(PlayerSettings {});
    player.can_play = false;
    set_rounds_played(player, 10);
    player.reset();
    EXPECT_TRUE(player.can_play);
    EXPECT_EQ(rounds_played(player), 0);
}

TEST_F(PlayerTest, set_human)
{
    Player player = Player::black(PlayerSettings {});
    EXPECT_TRUE(player.human());

    player.set_computer();
    EXPECT_TRUE(player.computer());
    EXPECT_FALSE(player.human());

    player.set_human();
    EXPECT_TRUE(player.human());
    EXPECT_FALSE(player.computer());
}

TEST_F(PlayerTest, player_type_string)
{
    Player player = Player::black(PlayerSettings {});
    EXPECT_EQ(player.type_string(), "Human   ");

    player.set_computer();
    EXPECT_EQ(player.type_string(), "Computer");
}

}  // namespace othello
